### PR TITLE
SLCORE-126 Introduce submonitor in the core to track subtask progress

### PR DESCRIPTION
**Pull request by henryju**

### CODE REVIEW

File: core/src/main/java/org/sonarsource/sonarlint/core/util/ProgressWrapper.java [hunk 1]
```diff
@@ -44,19 +56,23 @@ public void checkCancel() {
...
   public void setProgressAndCheckCancel(String msg, float fraction) {
     checkCancel();
     handler.setMessage(msg);
-    handler.setFraction(fraction);
+    setFraction(fraction);
   }
-  
+
+  private void setFraction(float fraction) {
+    handler.setFraction(offset + fraction * factor);
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/util/ProgressWrapper.java]: CRITICAL Use a "double" or "BigDecimal" instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2164)

File: core/src/main/java/org/sonarsource/sonarlint/core/util/ProgressWrapper.java [hunk 2]
```diff
@@ -19,22 +19,34 @@
...
       this.handler = handler;
     }
   }
 
+  public ProgressWrapper(@Nullable ProgressMonitor handler) {
+    this(0.0f, 1.0f, handler);
+  }
+
+  public ProgressWrapper subProgress(float fromFraction, float toFraction) {
+    return new ProgressWrapper(offset + fromFraction * factor, (toFraction - fromFraction) * factor, handler);
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/util/ProgressWrapper.java]: CRITICAL Use a "double" or "BigDecimal" instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2164)
CRITICAL Use a "double" or "BigDecimal" instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2164)

File: core/src/main/java/org/sonarsource/sonarlint/core/container/connected/SonarLintWsClient.java [hunk 3]
```diff
@@ -171,20 +171,22 @@ public String getOrganizationKey() {
...
         List<F> items = itemExtractor.apply(protoBufResponse);
         for (F item : items) {
           itemConsumer.accept(item);
+          loaded++;
         }
         boolean isEmpty = items.isEmpty();
         Paging paging = getPaging.apply(protoBufResponse);
         // SONAR-9150 Some WS used to miss the paging information, so iterate until response is empty
         stop = isEmpty || (paging.getTotal() > 0 && page * PAGE_SIZE >= paging.getTotal());
+        progress.setProgressAndCheckCancel("Loading page " + page, loaded / (float) paging.getTotal());
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/container/connected/SonarLintWsClient.java]: CRITICAL Use a "double" or "BigDecimal" instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2164)

File: core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/RulesDownloader.java [hunk 4]
```diff
@@ -65,20 +66,21 @@ public void fetchRulesTo(Path destDir) {
...
+      page++;
       SearchResponse response = loadFromStream(wsClient.get(getUrl(page, pageSize)));
       readPage(rulesBuilder, activeRulesBuildersByQProfile, response);
       loaded += response.getPs();
 
       if (response.getTotal() <= loaded) {
         break;
       }
-      page++;
+      progress.setProgressAndCheckCancel("Loading page " + page, loaded / (float) response.getTotal());
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/RulesDownloader.java]: CRITICAL Use a "double" or "BigDecimal" instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2164)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues
* ![CRITICAL][CRITICAL] 5 critical

Watch the comments in this conversation to review them.

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'

File: core/src/main/java/org/sonarsource/sonarlint/core/util/ProgressWrapper.java (hunk 2 above)
Comment (dbmeneses) [inline core/src/main/java/org/sonarsource/sonarlint/core/util/ProgressWrapper.java]: @henryju woudln't it be better for the subProgress to have a part of the message that is kept? The code to where the subProgress is passed might not know the context in which it is executed. Example, `subProgress("Downloading QPs", 0.4f, 0.6f); ` and then when calling `subProgress.setProgress("page 1/10", 0.1f)`, we would have the message "Download QPs (page 1/10)".

Comment (henryju) [inline core/src/main/java/org/sonarsource/sonarlint/core/util/ProgressWrapper.java]: Indeed, but what if there are more than 1 sub level? Also, I wonder if we should not support updating the progress without changing the message.

Comment (dbmeneses) [inline core/src/main/java/org/sonarsource/sonarlint/core/util/ProgressWrapper.java]: agree, that should be possible too.

**Pull request merged**